lib/Targets/AIETargetCDO.cpp

```diff
@@ -0,0 +1,1084 @@
+      << "\n"
+      << "/***************************** Includes "
+         "*********************************/\n"
+      << "//#include <fstream>\n"
+      << "extern \"C\"\n"
+      << "{\n"
+      << "  #include <xaiengine.h>\n"
+      << "}\n"
+      << "//#include \"adf/adf_api/AIEControlConfig.h\"\n\n"
+      << "#define __mlir_aie_try(x) x\n"
+      << "static XAie_DmaDimDesc *__mlir_aie_alloc_dim_desc(size_t ndims) {\n"
+      << "  XAie_DmaDimDesc *ret = NULL;\n"
+      << "  ret = (XAie_DmaDimDesc *)calloc(sizeof(XAie_DmaDimDesc), ndims);\n"
+      << "  if(NULL == ret) {\n"
+      << "    fprintf(stderr, \"Allocating DmaDimDesc failed.\\n\");\n"
+      << "  }\n"
+      << "  return ret;\n"
+      << "}\n";
+  return str;
+}
+
+static std::string cdoGenFileFooter() {
+  std::string str;
+  llvm::raw_string_ostream rss(str);
+  rss << "\n\n  class InitializeAIEControl\n"
+      << "  {\n"
+      << "  public:\n"
+      << "    InitializeAIEControl()\n"
+      << "    {\n"
+      << "      XAie_SetupConfig(ConfigPtr, HW_GEN, XAIE_BASE_ADDR, "
+         "XAIE_COL_SHIFT,\n"
+      << "                       XAIE_ROW_SHIFT, XAIE_NUM_COLS, "
+         "XAIE_NUM_ROWS,\n"
+      << "                       XAIE_SHIM_ROW, XAIE_MEM_TILE_ROW_START,\n"
+      << "                       XAIE_MEM_TILE_NUM_ROWS, "
+         "XAIE_AIE_TILE_ROW_START,\n"
+      << "                       XAIE_AIE_TILE_NUM_ROWS);\n"
+      << "\n"
+      << "      XAie_SetupPartitionConfig(&DevInst, XAIE_PARTITION_BASE_ADDR, "
+         "1, "
+      << numColsStr() << ");\n"
+      << "\n"
+      << "      XAie_CfgInitialize(&(DevInst), &ConfigPtr);\n"
+      << "\n"
+      << "#if defined(__AIESIM__)\n"
+      << "#if defined(__CDO__)\n"
+      << "      XAie_SetIOBackend(&(DevInst), XAIE_IO_BACKEND_CDO); // Set "
+         "aiengine driver library to run for CDO Mode\n"
+      << "      XAie_UpdateNpiAddr(&(DevInst), 0x0);\n"
+      << "#else\n"
+      << "      //AIE driver currently error out XAie_UpdateNpiAddr for "
+         "AIESIM\n"
+      << "#endif\n"
+      << "#else\n"
+      << "      XAie_UpdateNpiAddr(&(DevInst), 0x0);\n"
+      << "#endif\n"
+      << "\n"
+      << "#if defined(__AIESIM__) && !defined(__CDO__)\n"
+      << "      XAie_TurnEccOff(&DevInst);\n"
+      << "#endif\n"
+      << "\n"
+      << "#if defined(__AIESIM__) && !defined(__CDO__)\n"
+      << "      extern unsigned ess_debug;\n"
+      << "#else\n"
+      << "      unsigned ess_debug = false;\n"
+      << "#endif\n"
+      << "\n"
+      << "#ifdef __EXCLUDE_PL_CONTROL__\n"
+      << "      bool exclude_pl_control = true;\n"
+      << "#else\n"
+      << "      bool exclude_pl_control = false;\n"
+      << "#endif\n"
+      << "\n"
+      << "#ifdef __CDO__\n"
+      << "      int trace_config_stream_option = 2;\n"
+      << "#else\n"
+      << "      int trace_config_stream_option = 0;\n"
+      << "#endif\n"
+      << "    }\n"
+      << "  } initAIEControl;\n\n";
+  return str;
```

Comment:
At the I wonder whether it would not be easier to just provide some `.h` files and a run-time library instead of generating all this, for example if we want to call this from another program or runtime, to mix-and-match say AIE + GPU, potentially from other vendors.

Reply:
There's some code like you suggest checked into a new `data` area, but it would be better to delete that and the code you highlight. I think the preference is to generate the configuration bits directly and remove the dependence on the Vitis library. @jgmelber has already started some work down that path.